benchmark/utils/dpcpp_linops.dp.cpp

```diff
@@ -0,0 +1,195 @@
+
+
+#include "benchmark/utils/sparselib_linops.hpp"
+#include "benchmark/utils/types.hpp"
+
+
+class onemkl_csr {};
+class onemkl_optimized_csr {};
+
+
+namespace detail {
+
+
+inline oneapi::mkl::sparse::matrix_handle_t create_mat_handle()
+{
+    oneapi::mkl::sparse::matrix_handle_t mat_handle;
+    oneapi::mkl::sparse::init_matrix_handle(&mat_handle);
+    return mat_handle;
+}
+
+
+class OnemklBase : public gko::LinOp {
+public:
+    oneapi::mkl::sparse::matrix_handle_t get_mat_handle() const
+    {
+        return this->mat_handle_.get();
+    }
+
+    const gko::DpcppExecutor* get_device_exec() const
+    {
+        return device_exec_.get();
+    }
+
+protected:
+    void apply_impl(const gko::LinOp*, const gko::LinOp*, const gko::LinOp*,
+                    gko::LinOp*) const override
+    {
+        GKO_NOT_IMPLEMENTED;
+    }
+
+    void initialize_mat_handle()
+    {
+        // TODO: weird
+        mat_handle_ = handle_manager<oneapi::mkl::sparse::matrix_handle>(
+            create_mat_handle(),
+            [](oneapi::mkl::sparse::matrix_handle_t mat_handle) {
+                oneapi::mkl::sparse::release_matrix_handle(&mat_handle);
+            });
+    }
+
+    OnemklBase(std::shared_ptr<const gko::Executor> exec,
+               const gko::dim<2>& size = gko::dim<2>{})
+        : gko::LinOp(exec, size)
+    {
+        device_exec_ =
+            std::dynamic_pointer_cast<const gko::DpcppExecutor>(exec);
+        if (device_exec_ == nullptr) {
+            GKO_NOT_IMPLEMENTED;
+        }
+        this->initialize_mat_handle();
+    }
+
+    ~OnemklBase() = default;
+
+    OnemklBase(const OnemklBase& other) = delete;
+
+    OnemklBase& operator=(const OnemklBase& other)
+    {
+        if (this != &other) {
+            gko::LinOp::operator=(other);
+            this->device_exec_ = other.device_exec_;
+            this->initialize_mat_handle();
+        }
+        return *this;
+    }
+
+private:
+    std::shared_ptr<const gko::DpcppExecutor> device_exec_;
+    template <typename T>
+    using handle_manager = std::unique_ptr<T, std::function<void(T*)>>;
+    handle_manager<oneapi::mkl::sparse::matrix_handle> mat_handle_;
+};
+
+
+template <bool optimized = false, typename ValueType = gko::default_precision,
+          typename IndexType = gko::int32>
+class OnemklCsr
+    : public gko::EnableLinOp<OnemklCsr<optimized, ValueType, IndexType>,
+                              OnemklBase>,
+      public gko::EnableCreateMethod<
+          OnemklCsr<optimized, ValueType, IndexType>>,
+      public gko::ReadableFromMatrixData<ValueType, IndexType> {
+    friend class gko::EnableCreateMethod<OnemklCsr>;
+    friend class gko::EnablePolymorphicObject<OnemklCsr, OnemklBase>;
+
+public:
+    using Csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+    void read(const mat_data& data) override
+    {
+        csr_->read(data);
+        this->set_size(gko::dim<2>{csr_->get_size()});
```

Comment:
Couldn't you do it without creating a new `gko::dim<2>`?
```suggestion
        this->set_size(csr_->get_size());
```